### PR TITLE
Add ws string init from str

### PR DESCRIPTION
**Pull request by xVin**
targets #350

### CODE REVIEW

File: src/objects/string.c [hunk 1]
```diff
@@ -90,6 +90,34 @@ ws_string_init(
...
+
+    ws_object_unlock(&other->obj);
+
+    if (self->str) {
+        ws_object_unlock(&self->obj);
+        return true;
+    }
+
+    ws_object_unlock(&self->obj);
+    return false;
```
Comment (matthiasbeyer) [inline src/objects/string.c]: You can put the lines 112 to 118 into two lines:

``` C
ws_object_unlick(&self->obj);
return !!self->str;
```

File: src/objects/string.h [hunk 2]
```diff
@@ -62,7 +62,7 @@ extern ws_object_type_id WS_OBJECT_TYPE_ID_STRING;
  * @memberof ws_string
  *
  * @return true if ws_string object was successfully initialized,
- * NULL on failire
+ * NULL on failure
```
Comment (matthiasbeyer) [inline src/objects/string.h]: This is false anyway, a boolean cannot be `NULL`.

**Comment by xVin** [comment]
I added another commit to avoid code duplication.

File: src/objects/string.c [hunk 3]
```diff
@@ -90,6 +90,29 @@ ws_string_init(
...
+    if (!self || !other) {
+        return false;
+    }
+
+    ws_object_lock_write(&self->obj);
+    ws_object_lock_read(&other->obj);
+
+    self->charcount = other->charcount;
+
+    self->str = realloc(self->str, (self->charcount + 1) * sizeof(*self->str));
```
Comment (matthiasbeyer) [inline src/objects/string.c]: You lose `self-str` here when `realloc()` returns NULL.

File: src/objects/string.c [hunk 4]
```diff
@@ -90,6 +90,29 @@ ws_string_init(
...
+    self->str = u_strcpy(self->str, other->str);
```
Comment (matthiasbeyer) [inline src/objects/string.c:116]: Same problem here.

**Comment by matthiasbeyer** [comment]
ACK.

**Comment by neithernut** [comment]
ACK.

**Comment by matthiasbeyer** [comment]
Jup.

**Pull request merged**